File: Python/fsrs.py
import math
import random
from dataclasses import dataclass, replace
from datetime import timedelta
from enum import Enum
from typing import List


class Rating(Enum):
    AGAIN = 1
    HARD = 2
    GOOD = 3
    EASY = 4


class State(Enum):
    NEW = 0
    LEARNING = 1
    REVIEW = 2
    RELEARNING = 3


@dataclass
class Card:
    card_id: int
    interval: timedelta = timedelta(0)
    state: State = State.NEW
    step: int = 0
    stability: float = 0.0
    difficulty: float = 0.0


@dataclass(frozen=True)
class SchedulerConfig:
    w: List[float]
    desired_retention: float
    learning_steps: List[timedelta]
    relearning_steps: List[timedelta]
    maximum_interval: int
    enable_fuzzing: bool


DEFAULT_CONFIG = SchedulerConfig(
    w=[0.212, 1.2931, 2.3065, 8.2956, 6.4133, 0.8334, 3.0194, 0.001, 1.8722,
       0.1666, 0.796, 1.4835, 0.0614, 0.2629, 1.6483, 0.6014, 1.8729, 0.5425,
       0.0912, 0.0658, 0.1542],
    desired_retention=0.9,
    learning_steps=[timedelta(minutes=1), timedelta(minutes=10)],
    relearning_steps=[timedelta(minutes=10)],
    maximum_interval=36500,
    enable_fuzzing=True,
)

MIN_DIFFICULTY = 1.0
MAX_DIFFICULTY = 10.0
STABILITY_MIN = 0.001


def _clamp_difficulty(difficulty: float) -> float:
    return max(MIN_DIFFICULTY, min(difficulty, MAX_DIFFICULTY))


def _clamp_stability(stability: float) -> float:
    return max(stability, STABILITY_MIN)


def _raw_initial_difficulty(w: List[float], rating: Rating) -> float:
    return w[4] - (math.exp(w[5] * (rating.value - 1.0))) + 1.0


def initial_stability(w: List[float], rating: Rating) -> float:
    return _clamp_stability(w[rating.value - 1])


def initial_difficulty(w: List[float], rating: Rating) -> float:
    return _clamp_difficulty(_raw_initial_difficulty(w, rating))


def next_interval(factor: float, retention: float, decay: float,
                  max_interval: int, stability: float) -> timedelta:
    interval_days = (stability / factor) * (retention ** (1.0 / decay) - 1.0)
    return timedelta(days=min(max(1, round(interval_days)), max_interval))


def short_term_stability(w: List[float], stability: float, rating: Rating) -> float:
    increase = math.exp(w[17] * (rating.value - 3.0 + w[18])) * (stability ** -w[19])
    final_increase = max(increase, 1.0) if rating in [Rating.GOOD, Rating.EASY] else increase
    return _clamp_stability(stability * final_increase)


def next_difficulty(w: List[float], difficulty: float, rating: Rating) -> float:
    delta_difficulty = -w[6] * (rating.value - 3.0)
    damped_delta = ((MAX_DIFFICULTY - difficulty) * delta_difficulty /
                    (MAX_DIFFICULTY - MIN_DIFFICULTY))
    initial_easy_difficulty = _raw_initial_difficulty(w, Rating.EASY)
    new_difficulty = (w[7] * initial_easy_difficulty + (1.0 - w[7]) * (difficulty + damped_delta))
    return _clamp_difficulty(new_difficulty)


def _next_forget_stability(w: List[float], difficulty: float, stability: float, retrievability: float) -> float:
    long_term = (w[11] * (difficulty ** -w[12]) *
                 (((stability + 1.0) ** w[13]) - 1.0) *
                 (math.exp((1.0 - retrievability) * w[14])))
    short_term = stability / (math.exp(w[17] * w[18]))
    return min(long_term, short_term)


def _next_recall_stability(w: List[float], difficulty: float, stability: float,
                           retrievability: float, rating: Rating) -> float:
    hard_penalty = w[15] if rating == Rating.HARD else 1.0
    easy_bonus = w[16] if rating == Rating.EASY else 1.0
    return (stability *
            (1.0 + math.exp(w[8]) * (11.0 - difficulty) *
             (stability ** -w[9]) *
             (math.exp((1.0 - retrievability) * w[10]) - 1.0) *
             hard_penalty * easy_bonus))


def next_stability(w: List[float], difficulty: float, stability: float,
                   retrievability: float, rating: Rating) -> float:
    if rating == Rating.AGAIN:
        new_stability = _next_forget_stability(w, difficulty, stability, retrievability)
    else:
        new_stability = _next_recall_stability(w, difficulty, stability, retrievability, rating)
    return _clamp_stability(new_stability)


@dataclass(frozen=True)
class _FuzzRange:
    start: float
    end: float
    factor: float


class Scheduler:
    def __init__(self, config: SchedulerConfig = DEFAULT_CONFIG):
        self.config = config
        self.w = self._check_and_fill_parameters(config.w)
        self.decay = -self.w[20]
        self.factor = 0.9 ** (1.0 / self.decay) - 1.0
        self.rand = random.Random()
        self._fuzz_ranges = (
            _FuzzRange(start=2.5, end=7.0, factor=0.15),
            _FuzzRange(start=7.0, end=20.0, factor=0.1),
            _FuzzRange(start=20.0, end=float('inf'), factor=0.05),
        )

    def _check_and_fill_parameters(self, w: List[float]) -> List[float]:
        if any(not math.isfinite(val) for val in w):
            raise ValueError("Invalid parameters: contains non-finite values.")
        match len(w):
            case 21:
                return w
            case 19:
                return w + [0.0, 0.5]
            case 17:
                wn = list(w)
                wn[4] = wn[5] * 2.0 + wn[4]
                wn[5] = math.log(wn[5] * 3.0 + 1.0) / 3.0
                wn[6] = wn[6] + 0.5
                return wn + [0.0, 0.0, 0.0, 0.5]
            case _:
                raise ValueError(f"Invalid number of parameters: {len(w)}. Supported: 17, 19, or 21.")

    def _get_card_retrievability(self, card: Card, review_interval: timedelta) -> float:
        elapsed_days = max(0.0, review_interval.total_seconds() / 86400)
        return (1.0 + self.factor * elapsed_days / card.stability) ** self.decay

    def _calculate_next_reviewed_state(self, card: Card, rating: Rating, review_interval: timedelta):
        if card.state == State.NEW:
            card.stability = initial_stability(self.w, rating)
            card.difficulty = initial_difficulty(self.w, rating)
            card.state = State.LEARNING
            card.step = 0
        else:
            retrievability = self._get_card_retrievability(card, review_interval)

            new_difficulty = next_difficulty(self.w, card.difficulty, rating)
            new_stability = (short_term_stability(self.w, card.stability, rating)
                             if review_interval.total_seconds() < 86400
                             else next_stability(self.w, card.difficulty, card.stability, retrievability, rating))

            card.difficulty = new_difficulty
            card.stability = new_stability

    def _calculate_next_review_interval(self, stability: float) -> timedelta:
        return next_interval(
            self.factor, self.config.desired_retention, self.decay,
            self.config.maximum_interval, stability
        )

    def _to_review_state(self, card: Card):
        card.state = State.REVIEW
        card.step = 0
        card.interval = self._calculate_next_review_interval(card.stability)

    def _hard_interval_step(self, card: Card, steps: List[timedelta]) -> timedelta:
        if card.step == 0 and len(steps) == 1:
            return timedelta(minutes=steps[0].total_seconds() / 60 * 1.5)
        if card.step == 0 and len(steps) > 1:
            return timedelta(minutes=(steps[0].total_seconds() / 60 + steps[1].total_seconds() / 60) / 2.0)
        return steps[card.step]

    def _handle_steps(self, card: Card, rating: Rating, steps: List[timedelta]):
        if not steps:
            self._to_review_state(card)
            return
        match rating:
            case Rating.AGAIN:
                card.state = State.LEARNING
                card.step = 0
                card.interval = steps[0]
            case Rating.HARD:
                card.interval = self._hard_interval_step(card, steps)
            case Rating.GOOD:
                next_step = card.step + 1
                if next_step >= len(steps):
                    self._to_review_state(card)
                else:
                    card.state = State.LEARNING
                    card.step = next_step
                    card.interval = steps[next_step]
            case Rating.EASY:
                self._to_review_state(card)

    def _determine_next_card_state(self, card: Card, rating: Rating):
        match card.state:
            case State.LEARNING:
                self._handle_steps(card, rating, self.config.learning_steps)
            case State.RELEARNING:
                self._handle_steps(card, rating, self.config.relearning_steps)
            case State.REVIEW:
                if rating == Rating.AGAIN and self.config.relearning_steps:
                    card.state = State.RELEARNING
                    card.step = 0
                    card.interval = self.config.relearning_steps[0]
                else:
                    self._to_review_state(card)

    def _get_fuzzed_interval(self, interval: timedelta) -> timedelta:
        interval_days = interval.total_seconds() / 86400
        if interval_days < 2.5:
            return interval
        delta = sum(r.factor * max(0.0, min(interval_days, r.end) - r.start) for r in self._fuzz_ranges)
        min_ivl = max(2, round(interval_days - delta))
        max_ivl = round(interval_days + delta)
        fuzzed_days = self.rand.randint(min_ivl, max_ivl)
        return timedelta(days=min(fuzzed_days, self.config.maximum_interval))

    def _apply_fuzzing(self, card: Card):
        if self.config.enable_fuzzing and card.state == State.REVIEW:
            card.interval = self._get_fuzzed_interval(card.interval)

    def review_card(self, card: Card, rating: Rating, review_interval: timedelta) -> None:
        self._calculate_next_reviewed_state(card, rating, review_interval)
        self._determine_next_card_state(card, rating)
        self._apply_fuzzing(card)

File: Elixir/fsrs.ex
defmodule Card do
  defstruct card_id: nil, interval: 0, phase: :new

  defmodule ReviewedCard do
    defstruct state: :learning, step: 0, stability: 0.0, difficulty: 0.0
  end

  defmodule Config do
    defstruct w: [
                0.212, 1.2931, 2.3065, 8.2956, 6.4133, 0.8334, 3.0194, 0.001, 1.8722, 0.1666,
                0.796, 1.4835, 0.0614, 0.2629, 1.6483, 0.6014, 1.8729, 0.5425, 0.0912,
                0.0658, 0.1542
              ],
              desired_retention: 0.9,
              learning_steps: [1 / 1440, 10 / 1440],
              relearning_steps: [10 / 1440],
              maximum_interval: 36_500,
              enable_fuzzing: true
  end
end

defmodule Algorithm do
  alias Card.ReviewedCard

  @min_difficulty 1.0
  @max_difficulty 10.0
  @stability_min 0.001

  defp rating_to_value(:again), do: 1
  defp rating_to_value(:hard), do: 2
  defp rating_to_value(:good), do: 3
  defp rating_to_value(:easy), do: 4

  defp clamp_difficulty(difficulty), do: max(@min_difficulty, difficulty) |> min(@max_difficulty)
  defp clamp_stability(stability), do: max(stability, @stability_min)

  defp raw_initial_difficulty(w, rating) do
    rating_value = rating_to_value(rating)
    elem(w, 4) - :math.exp(elem(w, 5) * (rating_value - 1.0)) + 1.0
  end

  def initial_stability(w, rating) do
    w
    |> elem(rating_to_value(rating) - 1)
    |> clamp_stability()
  end

  def initial_difficulty(w, rating) do
    w
    |> raw_initial_difficulty(rating)
    |> clamp_difficulty()
  end

  def next_interval(factor, retention, decay, max_interval, stability) do
    (stability / factor) * (retention**(1.0 / decay) - 1.0)
    |> round()
    |> max(1)
    |> min(max_interval)
    |> Kernel.to_float()
  end

  def next_stability(w, data, retrievability, :again) do
    long_term =
      elem(w, 11) * data.difficulty**-elem(w, 12) *
        ((data.stability + 1.0)**elem(w, 13) - 1.0) *
        :math.exp((1.0 - retrievability) * elem(w, 14))

    short_term = data.stability / :math.exp(elem(w, 17) * elem(w, 18))

    min(long_term, short_term)
    |> clamp_stability()
  end

  def next_stability(w, data, retrievability, rating) do
    hard_penalty = if rating == :hard, do: elem(w, 15), else: 1.0
    easy_bonus = if rating == :easy, do: elem(w, 16), else: 1.0

    data.stability *
      (1.0 +
         :math.exp(elem(w, 8)) * (11.0 - data.difficulty) * data.stability**-elem(w, 9) *
           (:math.exp((1.0 - retrievability) * elem(w, 10)) - 1.0) * hard_penalty *
           easy_bonus)
    |> clamp_stability()
  end

  def next_difficulty(w, difficulty, rating) do
    rating_value = rating_to_value(rating)
    delta_difficulty = -(elem(w, 6) * (rating_value - 3.0))

    damped_delta =
      (@max_difficulty - difficulty) * delta_difficulty / (@max_difficulty - @min_difficulty)

    initial_easy_difficulty = raw_initial_difficulty(w, :easy)

    elem(w, 7) * initial_easy_difficulty +
      (1.0 - elem(w, 7)) * (difficulty + damped_delta)
    |> clamp_difficulty()
  end

  def short_term_stability(w, stability, rating) do
    rating_value = rating_to_value(rating)

    increase =
      :math.exp(elem(w, 17) * (rating_value - 3.0 + elem(w, 18))) *
        stability**-elem(w, 19)

    final_increase = if rating in [:good, :easy], do: max(increase, 1.0), else: increase

    stability * final_increase
    |> clamp_stability()
  end
end

defmodule Scheduler do
  alias Card
  alias Card.{ReviewedCard, Config}
  alias Algorithm

  defmodule State do
    defstruct config: nil, w: nil, decay: 0.0, factor: 0.0
  end

  @fsrs5_default_decay 0.5

  def create(config \\ %Config{}) do
    with {:ok, w} <- check_and_fill_parameters(config.w) do
      decay = -elem(w, 20)
      factor = 0.9**(1.0 / decay) - 1.0
      state = %State{config: config, w: w, decay: decay, factor: factor}
      {:ok, state}
    end
  end

  def review_card(scheduler_state, card, rating, review_interval \\ 0) do
    card
    |> calculate_next_reviewed_state(scheduler_state, rating, review_interval)
    |> determine_next_phase_and_interval(scheduler_state, rating)
    |> apply_fuzzing(scheduler_state)
    |> update_card(card)
  end

  def calculate_next_interval(scheduler_state, stability) do
    Algorithm.next_interval(
      scheduler_state.factor,
      scheduler_state.config.desired_retention,
      scheduler_state.decay,
      scheduler_state.config.maximum_interval,
      stability
    )
  end

  defp update_card({final_reviewed, final_interval}, original_card) do
    %{original_card | interval: final_interval, phase: final_reviewed}
  end

  defp check_and_fill_parameters(w_list) when is_list(w_list) do
    filled =
      case length(w_list) do
        17 ->
          w5 = Enum.at(w_list, 5)
          w_list
          |> List.update_at(4, &(&1 + w5 * 2.0))
          |> List.update_at(5, fn _ -> :math.log(w5 * 3.0 + 1.0) / 3.0 end)
          |> List.update_at(6, &(&1 + 0.5))
          |> Kernel.++([0.0, 0.0, 0.0, @fsrs5_default_decay])
        19 ->
          w_list ++ [0.0, @fsrs5_default_decay]
        21 ->
          w_list
        _ ->
          :error
      end
    if filled == :error,
      do: {:error, "Invalid number of parameters. Supported: 17, 19, or 21."},
      else:
        if Enum.any?(filled, fn val -> !is_float(val) or !Float.is_finite(val) end),
          do: {:error, "Invalid parameters: contains non-finite values."},
          else: {:ok, List.to_tuple(filled)}
  end

  defp calculate_next_reviewed_state(%Card{phase: :new}, scheduler, rating, _) do
    stability = Algorithm.initial_stability(scheduler.w, rating)
    difficulty = Algorithm.initial_difficulty(scheduler.w, rating)
    %ReviewedCard{state: :learning, step: 0, stability: stability, difficulty: difficulty}
  end

  defp calculate_next_reviewed_state(%Card{phase: data}, scheduler, rating, review_interval) do
    new_difficulty = Algorithm.next_difficulty(scheduler.w, data.difficulty, rating)
    new_stability =
      if review_interval < 1.0,
        do: Algorithm.short_term_stability(scheduler.w, data.stability, rating),
        else:
          get_card_retrievability(scheduler, data, review_interval)
          |> Algorithm.next_stability(scheduler.w, data, rating)
    %ReviewedCard{
      state: data.state,
      step: data.step,
      stability: new_stability,
      difficulty: new_difficulty
    }
  end

  defp get_card_retrievability(scheduler, card_data, review_interval) do
    elapsed_days = max(0.0, review_interval)
    (1.0 + scheduler.factor * elapsed_days / card_data.stability)**scheduler.decay
  end

  defp determine_next_phase_and_interval(
         %ReviewedCard{state: :learning} = reviewed,
         scheduler,
         rating
       ),
       do: handle_steps(scheduler, reviewed, rating, scheduler.config.learning_steps)

  defp determine_next_phase_and_interval(
         %ReviewedCard{state: :relearning} = reviewed,
         scheduler,
         rating
       ),
       do: handle_steps(scheduler, reviewed, rating, scheduler.config.relearning_steps)

  defp determine_next_phase_and_interval(
         %ReviewedCard{state: :review} = reviewed,
         scheduler,
         rating
       ),
       do: handle_review(scheduler, reviewed, rating)

  defp handle_review(scheduler, reviewed, :again) when scheduler.config.relearning_steps == [],
    do: to_review_state(scheduler, reviewed)

  defp handle_review(scheduler, reviewed, :again) do
    [h | _] = scheduler.config.relearning_steps
    {%{reviewed | state: :relearning, step: 0}, h}
  end

  defp handle_review(scheduler, reviewed, _), do: to_review_state(scheduler, reviewed)

  defp handle_steps(scheduler, reviewed, _, []), do: to_review_state(scheduler, reviewed)
  defp handle_steps(_, reviewed, :again, [h | _]), do: {%{reviewed | state: :learning, step: 0}, h}

  defp handle_steps(_, reviewed, :hard, steps) do
    interval =
      case {reviewed.step, steps} do
        {0, [s1]} -> s1 * 1.5
        {0, [s1, s2 | _]} -> (s1 + s2) / 2.0
        {step, _} -> Enum.at(steps, step)
      end
    {%{reviewed | state: reviewed.state}, interval}
  end

  defp handle_steps(scheduler, reviewed, :good, steps) do
    next_step = reviewed.step + 1
    handle_good_step(scheduler, reviewed, steps, next_step)
  end

  defp handle_steps(scheduler, reviewed, :easy, _), do: to_review_state(scheduler, reviewed)

  defp handle_good_step(scheduler, reviewed, steps, next_step) when next_step >= length(steps),
    do: to_review_state(scheduler, reviewed)

  defp handle_good_step(_, reviewed, steps, next_step) do
    {%{reviewed | state: :learning, step: next_step}, Enum.at(steps, next_step)}
  end

  defp to_review_state(scheduler, reviewed) do
    interval = calculate_next_interval(scheduler, reviewed.stability)
    {%{reviewed | state: :review, step: 0}, interval}
  end

  defp apply_fuzzing({%ReviewedCard{state: :review} = reviewed, interval}, scheduler)
       when scheduler.config.enable_fuzzing and interval >= 2.5 do
    fuzzed_interval = get_fuzzed_interval(scheduler.config.maximum_interval, interval)
    {reviewed, fuzzed_interval}
  end

  defp apply_fuzzing({reviewed, interval}, _) do
    {reviewed, interval}
  end

  defp get_fuzzed_interval(max_interval, interval) do
    fuzz_ranges = [
      %{start: 2.5, end: 7.0, factor: 0.15},
      %{start: 7.0, end: 20.0, factor: 0.1},
      %{start: 20.0, end: :infinity, factor: 0.05}
    ]
    delta =
      Enum.reduce(fuzz_ranges, 0.0, fn range, acc ->
        range_end = if range.end == :infinity, do: interval, else: range.end
        acc + range.factor * max(0.0, min(interval, range_end) - range.start)
      end)
    min_ivl = round(interval - delta) |> max(2)
    max_ivl = round(interval + delta)
    if min_ivl >= max_ivl,
      do: min(min_ivl, max_interval) |> Kernel.to_float(),
      else:
        :rand.uniform(max_ivl - min_ivl + 1) + min_ivl - 1
        |> min(max_interval)
        |> Kernel.to_float()
  end
end
